webhook/server.js

```diff
@@ -0,0 +1,377 @@
+                const room = event.payload.room
+                const channel = await parent.getOrCreateRoom((await parent.getChatKitRoom(room.id, room.created_by_id)))
+                const members = [];
+                // ensure usersCache exists
+                event.payload.users.forEach(async function (u) {
+                    members.push(parent.sanitizeUserId(u.id))
+                    await parent.createStreamUser(u)
+                })
+                await channel.removeMembers(members)
+            },
+            "v1.rooms_deleted": async function (event) {
+                //this is unfortunate, we are probably creating the room right before deleting it
+                event.payload.room_ids.forEach(async function (roomId) {
+                    // todo implement me
+                })
+            }
+        }
+    }
+
+    async getChatKitRoom(roomID, user) {
+        const cachedRoom = this.roomsCache.get(roomID)
+        if (cachedRoom) {
+            return cachedRoom
+        }
+        const room = await this.chatKitClient().getRoom({
+            userId: user,
+            roomId: roomID,
+        })
+        this.roomsCache.set(roomID, room)
+        return room
+    }
+
+    async ensureMembersExists(ids) {
+
+        const missingIds = []
+        const parent = this
+
+        for (let i = 0; i < ids.length; i++) {
+            const user = this.usersCache.get(ids[i])
+            if (!user) {
+                missingIds.push(ids[i])
+            }
+        }
+
+        // get missing users from cache
+        if (missingIds.length > 0) {
+            const loadedUsers = await this.chatKitClient().getUsersById({
+                userIds: missingIds
+            })
+
+            const users = []
+            loadedUsers.forEach(function (u) {
+                users.push(parent.toStreamUser(u))
+            })
+            await this.streamClient().updateUsers(users)
+
+            users.forEach(function (u) {
+                // update LRU cache
+                parent.usersCache.set(u.id, u)
+            })
+        }
+    }
+
+    // sanitizeUserId transform a chatKit user id to a suitable stream user id
+    sanitizeUserId(id) {
+        // todo improve me
+        return id.replace(':', '_')
+    }
+
+    // converts a chatKit user to a stream user
+    async createStreamUser(chatKitUser) {
+        await this.streamClient().updateUser(this.toStreamUser(chatKitUser))
+    }
+
+    // converts a chatKit user user to a a stream user
+    toStreamUser(chatKitUser) {
+        return {
+            id: this.sanitizeUserId(chatKitUser.id),
+            image: chatKitUser.avatar_url,// todo omit keys with empty values
+            name: chatKitUser.name,
+            ...chatKitUser.custom_data,
+        };
+    }
+
+    // converts a chatKit room to a stream channel
+    async getOrCreateRoom(room) {
+        let type = 'livestream'
+        if (room.private) {
+            type = 'messaging'
+        }
```

Comment:
A functor to determine channelType would be ideal here. For us, as an example, we have a custom field on the room that will map to which channelType it'll be.